### ISSUE TITLE
Github registration requests personal information.

### ISSUE THREAD

**Issue by 1337cookie** [userstyles-world/userstyles.world#61]
Github registration requests access to "Personal user data". Github's documentation isn't super specific about what information that is. 
[From the docs:](https://docs.github.com/en/github/authenticating-to-github/keeping-your-account-and-data-secure/authorizing-oauth-apps#types-of-requested-data#:~:text=Personal%20user%20data)

> User data includes information found in your user profile, like your name, e-mail address, and location.

Is this information necessary for userstyles-world to have access to? Could it be dialed back? I would prefer not to have to trust more third parties than necessary to handle my personal information.

Screenshot of authentication page:
![image](https://user-images.githubusercontent.com/15826769/120698954-db50a000-c503-11eb-88f0-8e8ad187111d.png)

**Comment by Gusted** [comment]
Hi @1337cookie!

Thank you for question, we might consider this to add this as a FAQ or something. @vednoc?

> Is this information necessary for userstyles-world to have access to?

Yes, but no, but actually yes.

Sounds confusing right, because it is.
So let take a step back and ask the question "Why do we need it?"

Currently we are using this integration with github/gitlab/codeberg to receiver's the username of the authorized user. And we will be later on adding emails to communicate important messages with our users. And because we want to make sure to have an authorized way to check if the user is actually this user on the said platform.

We are requesting the "read:user" permission which already eliminates to "write" permission to modify the data, because we just care to read certain data. https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps:
![image](https://user-images.githubusercontent.com/25481501/120700536-4cab3700-c4a1-11eb-9e4c-3c18cb8dba40.png)

At the moment we receive an token that we can use to interact with github's endpoints on behalf of the user we are requesting the `/user` endpoint which you can read more about: https://docs.github.com/en/rest/reference/users#get-the-authenticated-user and yes at this moment it will send `Response with public and private profile information` to us. At the moment I don't see another way to make sure we receive the display name other then this. Because most if not all other endpoints that could give us this insight REQUIRES to know the display_name already, which we quite frankly are trying to figure out which is.

Now when this data is received we are storing(temporarily as variable), but within this variable it's just a `struct/object` that holds the username. And the token that we actually could use indefinably until revoked by user isn't getting stored and just freed out of memory.

> Could it be dialed back?

Unless if the response given at https://docs.github.com/en/rest/reference/users#get-the-authenticated-user is somehow very personal because it very limits the private information given to us, no then not.



> I would prefer not to have to trust more third parties than necessary to handle my personal information.

I understand, I get that. You always could choose not to use github or anything and simple use the email/password way. This integration with other parties was to make it more convince to register an account.

Regards,
Gusted

**Comment by vednoc** [comment]
In short, a better way to get user's username and e-mail doesn't exist, so we must do it that way.

The data that we get from GitHub (and/or other OAuth providers) is not logged anywhere, and we just use the two aforementioned fields in order to create your account. It also allows us to contact you directly via e-mail in case that is ever necessary.

> we might consider this to add this as a FAQ or something

Great idea. I'll include a short form and a link to this issue in the work-in-progress FAQ page.

**Comment by 1337cookie** [comment]
> > I would prefer not to have to trust more third parties than necessary to handle my personal information.
> 
> I understand, I get that. You always could choose not to use github or anything and simple use the email/password way. This integration with other parties was to make it more convince to register an account.

I wouldn't sign up with email/password so you're correct this is the path that would make me want to sign up. My last email account has been linked with over 100 data breaches on various websites. Not trusting as many services as possible to safely store passwords is good but but now I am asked for personal data which will include my email address which was half the reason to use a 3rd party login. 

From my understanding of the scopes you can use no scope and still validate the account but you wont have access to my email. Example: 

![image](https://user-images.githubusercontent.com/15826769/120704722-12767f80-c50b-11eb-98aa-2125f5955645.png)


I will revise my question to:
Could the service operate without access to my email address?

**Comment by Gusted** [comment]
> From my understanding of the scopes you can use no scope and still validate the account but you wont have access to my email.

Yeah and no. From my understanding and playing around with the endpoints a while, we will have access to public information, INCLUDING the display_name which we want. 

However we are requiring email as an mandatory field. For 2 main reasons
- Legals, we should, when needed to contact a user by their provided email which is verified either by us or the third party. 
- Announcement, well this isn't the case now. We in the future could announce certain big chances, ex. privacy policy update or Terms of Services.

I'm not sure if we should allow user accounts without confirmation of an email address as that opens to a fast access of spam accounts creations, which is a fall-pit for certain services that allow this especially as we don't want to force down some captcha(which includes more work and things I don't like) and that collects a lot lot lot more information about you.

So, could this service operator without access to your email: No not really, it's being used as a verification proof and for the 2 other reasons mentioned.

**Comment by 1337cookie** [comment]
> I'm not sure if we should allow user accounts without confirmation of an email address as that opens to a fast access of spam accounts creations, which is a fall-pit for certain services that allow this especially as we don't want to force down some captcha(which includes more work and things I don't like) and that collects a lot lot lot more information about you.

Captcha is useless, you can pay a service to have people complete them for you (internet search: captcha jobs). Github will only allow OAuth and Github apps connections from accounts with verified email and blocks those actions from accounts flagged as spam so you get that additional verification for free vs rolling your own verification.
In the event your database was compromised an attacker would have my Github username and email address which could allow a more sophisticated phishing attack which my primate brain is vulnerable to.
My preference is toward not receiving correspondence from many services (No disrespect, this project is awesome hence my engagement). If I'm actively engaged with the service an internal message system or frontpage can notify me and If there was some kind of data breach notifying me wouldn't be necessary since I don't have any personal information stored.
Thanks for taking the time to respond and discuss.

**Comment by Gusted** [comment]
> Captcha is useless, you can pay a service to have people complete them for you (internet search: captcha jobs).

Yeah I know about them, now I'm not sure but pretty sure you still have to load the page with the captcha on them and they will still be loading and tracking you which I don't trust any third-party with to handle this. 

> Github will only allow OAuth and Github apps connections from accounts with verified email and blocks those actions from accounts flagged as spam so you get that additional verification for free vs rolling your own verification.

Didn't know that. Not so sure about the other Gitlab and Codeberg.



> In the event your database was compromised an attacker would have my Github username and email address which could allow a more sophisticated phishing attack which my primate brain is vulnerable to.

Correct. Bit off-topic:

However in this case with certain users that also uses github/gitlab/codeberg for the intended purpose of having git repo's the email is likely already exposed due to how you will send "patches" to these services.
And playing the advocate of the devil, having you as an example the email is already likely exposed. And to not just show it here in this commit Il be sending a email to the email I found from your account, with the link of where I found it.


^^ And now on-topic again



>  If I'm actively engaged with the service an internal message system or frontpage can notify me and If there was some kind of data breach notifying me wouldn't be necessary since I don't have any personal information stored.

I'm not sure and **DISCLAMER: IANAL** those aren't legally binding of notifying the user. What if the user abandoned the site and had PII in our database of them. We don't want to track when's the last time the user logged into their account etc, so not reliable of when we simply can send a warning/info(which for someone also can be overlooked etc) and when to send a email.

Il be looking into removing the `read:user` and have it limited to the `user:email` that we currently have specified for Github.

Regards,
Gusted

**Comment by Gusted** [comment]
https://github.com/userstyles-world/userstyles.world/commit/25a5ce3083985a972e59197117fccb8f5c572838 Will be included in the next server update.
